samples/net/secure_mqtt_sensor_actuator/src/mqtt_client.c

```diff
@@ -0,0 +1,562 @@
+
+	rc = mqtt_subscribe(client, &sub_list);
+	if (rc != 0) {
+		LOG_ERR("MQTT Subscribe failed [%d]", rc);
+	}
+
+	return rc;
+}
+
+/** Process incoming MQTT data and keep the connection alive*/
+int app_mqtt_process(struct mqtt_client *client)
+{
+	int rc;
+
+	rc = poll_mqtt_socket(client, mqtt_keepalive_time_left(client));
+
+	if (rc != 0) {
+		if (fds[0].revents & ZSOCK_POLLIN) {
+			/* MQTT data received */
+			rc = mqtt_input(client);
+			if (rc != 0) {
+				LOG_ERR("MQTT Input failed [%d]", rc);
+				return rc;
+			}
+			/* Socket error */
+			if (fds[0].revents & (ZSOCK_POLLHUP | ZSOCK_POLLERR)) {
+				LOG_ERR("MQTT socket closed / error");
+				return -ENOTCONN;
+			}
+		}
+	} else {
+		/* Socket poll timed out, time to call mqtt_live() */
+		rc = mqtt_live(client);
+		if (rc != 0) {
+			LOG_ERR("MQTT Live failed [%d]", rc);
+			return rc;
+		}
+	}
+
+	return 0;
+}
+
+/** Connect to MQTT broker
+ *  Does not return until a connection is made to the MQTT broker.
+ */
+void app_mqtt_connect(struct mqtt_client *client)
+{
+	int rc = 0;
+
+	mqtt_connected = false;
+
+	/* Block until MQTT CONNACK event callback occurs */
+	while (!mqtt_connected) {
+		rc = mqtt_connect(client);
+		if (rc != 0) {
+			LOG_ERR("MQTT Connect failed [%d]", rc);
+			k_msleep(MSECS_WAIT_RECONNECT);
+			continue;
+		}
+
+		/* Poll MQTT socket for response */
+		rc = poll_mqtt_socket(client, MSECS_NET_POLL_TIMEOUT);
+		if (rc > 0) {
+			mqtt_input(client);
+		}
+
+		if (!mqtt_connected) {
+			mqtt_abort(client);
+		}
+	}
+}
+
+/** MQTT thread main loop.
+ *  Subscribes to MQTT topics and continuously
+ *  processes incoming data while the connection is active.
+ */
+void app_mqtt_run(struct mqtt_client *client)
+{
+	int rc;
+
+	/* Subscribe to MQTT topics */
+	app_mqtt_subscribe(client);
+
+	/* Thread will primarily remain in this loop */
+	while (mqtt_connected) {
+		rc = app_mqtt_process(client);
+		if (rc != 0) {
+			break;
+		}
+	}
+	/* Gracefully close connection */
+	mqtt_disconnect(client);
+}
+
+
+/** The system work queue is used to handle MQTT publishing.
+ *  A timer is set up to trigger a periodic interrupt, at which point
+ *  an MQTT publish will be queued in the system work queue.
+ *  Use CONFIG_MQTT_PUBLISH_INTERVAL to set the publish frequency.
+ */
+static void publish_work_handler(struct k_work *work)
+{
+	int rc;
+
+	rc = app_mqtt_publish(&client_ctx);
+	if (rc != 0) {
+		LOG_INF("MQTT Publish failed [%d]", rc);
+	}
+}
+
+K_WORK_DEFINE(publish_work, publish_work_handler);
+
+static void publish_timer_handler(struct k_timer *t)
+{
+	/* Queue MQTT publish on timer interrupt, if the connection is up */
+	if (mqtt_connected) {
+		k_work_submit(&publish_work);
+	}
+}
+
+K_TIMER_DEFINE(publish_timer, publish_timer_handler, NULL);
+
+/** Main thread entry */
+void mqtt_client_thread(void)
+{
+	/* Wait for network to come up */
+	while (k_sem_take(&net_conn_sem, K_MSEC(MSECS_NET_POLL_TIMEOUT)) != 0) {
```

Comment:
IMO this could pretty much continue on the main thread, no need to start a separate thread.

Reply:
Good point - was an early design decision that I never looked back on.
I moved the MQTT client functionlity to the main thread.
I created a new header file to expose the MQTT client APIs in mqtt_client.c
Increased main thread stack size from 1024 to 2048.